src/server.ts

```diff
@@ -0,0 +1,147 @@
+import {
+  REST,
+  Routes,
+  OAuth2Routes,
+  type RESTPutAPICurrentUserApplicationRoleConnectionJSONBody,
+  type RESTGetAPICurrentUserConnectionsResult,
+  type RESTGetAPIUserResult,
+  ConnectionService,
+  RESTPostOAuth2AccessTokenResult,
+} from 'discord.js';
+
+import Fastify, { RequestGenericInterface } from 'fastify';
+
+import { areContributors, contributorsStored, storeToken } from './storage';
+import config from './config';
+
+interface OAuth2Callback extends RequestGenericInterface {
+  Querystring: {
+    code: string;
+  };
+}
+
+const makeRestAPI = (accessToken: string | undefined) => {
+  const rest = new REST({
+    version: '10',
+    authPrefix: 'Bearer',
+  });
+  if (accessToken) rest.setToken(accessToken);
+  return rest;
+};
+
+// TODO: add state param
+const generateAuthorizeUrl = () => {
+  const url = new URL(OAuth2Routes.authorizationURL);
+  url.searchParams.append('client_id', config.discord.clientId);
+  url.searchParams.append('redirect_uri', config.discord.oauth2.redirectUri);
+  url.searchParams.append('response_type', 'code');
+  url.searchParams.append('scope', config.discord.oauth2.scope);
+  return url.toString();
+};
+
+const getTokensFromOAuth = async (rest: REST, code: string) => {
+  return (await rest.post(Routes.oauth2TokenExchange(), {
+    auth: false,
+    body: new URLSearchParams({
+      client_id: config.discord.clientId,
+      client_secret: config.discord.clientSecret,
+      code,
+      grant_type: 'authorization_code',
+      redirect_uri: config.discord.oauth2.redirectUri,
+      scope: config.discord.oauth2.scope,
+    }),
+    passThroughBody: true,
+    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
+  })) as RESTPostOAuth2AccessTokenResult;
+};
+
+const getDiscordProfile = async (rest: REST) => {
+  return (await rest.get(Routes.user())) as RESTGetAPIUserResult;
+};
+
+const getGitHubConnections = async (rest: REST) => {
+  const connections = (await rest.get(
+    Routes.userConnections()
+  )) as RESTGetAPICurrentUserConnectionsResult;
+
+  return connections.reduce((result, connection) => {
+    if (connection.type == ConnectionService.GitHub) result.push(connection.id);
+    return result;
+  }, [] as string[]);
```

Comment:
```suggestion
  return connections
    .filter(connection => connection.type === ConnectionService.GitHub)
    .map(connection => connection.id);
```

Reply:
I intentionally avoided this, as it would result in two iterations, instead of just one using reduce

Reply:
hm, i can't imagine any big performance impact